enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,205 @@
+---
+title: on-prem-configurable-lb-dns-stack
+authors:
+  - "@yboaron"
+reviewers:
+  - "@sttts"
+  - "@deads2k"
+  - "@celebdor"
+  - "@bcrochet"
+  - "@bnemec"
+approvers:
+  - "@"
+creation-date: 2020-02-11
+last-updated: 2020-02-11
+status: 
+see-also:
+replaces:
+superseded-by:
+
+---
+
+# On-prem: configurable DNS/LB self-hosted stack 
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+In the current implementation,  the self-hosted DNS/LB stack runs under the MCO umbrella by default in all on-prem platforms and there's no option to configure it. 
+Some customers with high production needs usually have their own external Load Balancing and DNS resolution. In addition, there are cases in which some traffic used to provide these self-hosted services is disallowed in the customer's network.                                                                                                                                                                                                                                        The outcome of such scenarios is that cluster's resources are spent on providing unused services and generating traffic against certain customers network design.
+
+This enhancement suggests a new method for providing the self-hosted stack.
+ 
+*Note:* as per the OCP plan to move API and Ingress to K8S SLB, this direction should be investigated deeper after MetalLb will be integrated.
+
+
+## Motivation
+
+Allow customers to adjust the self hosted DNS/LB stack according to the needs they really have instead of providing them in a generic way. 
+Additionally, detaching the self hosted stack configuration from the selected platform will allow new customers from UPI (platform: None) to leverage the self-hosted stack.
+
+
+### Goals
+
+- Ability to use the self-hosted stack in all on-prem platforms.
+- Ability to control the self-hosted stack configuration.
+- Retrieve the self-hosted stack status through K8S API.
+- Maintain a single code base/manifests for all the on-prem platforms.
+
+
+## Proposal
+
+The [Keepalived and infra DNS components](https://github.com/openshift/machine-config-operator/tree/master/manifests/baremetal) on the bootstrap node will keep running through the MCO.
+However, the services on masters and workers nodes will be managed by a new operator instead of static pods.
```

Comment:
we discussed this in Slack: for disaster recovery and cluster restart we will probably need to keep keepalived and haproxy run through static pods. We cannot rely on normal pods as these create cyclic dependencies: kubelet, kube-controller-manager and kube-scheduler need a working API VIP to run those pods.